### PR TITLE
[SPARK-32288][UI] Add exception summary for failed tasks in stage page

### PR DESCRIPTION
**Pull request by warrenzhu25**
### What changes were proposed in this pull request?
Add exception summary table for failed tasks in stage page.
Before:
image

After:
image

### Why are the changes needed?
When there're many task failure during one stage, it's hard to find failure pattern such as aggregation task failure by exception type and message. If we have such information, we can easily know which type of exception of failure is the root cause of stage failure.

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
Added UT in UISeleniumSuite

### CODE REVIEW

**Comment by Tagar** [comment]
Great improvement 

Would be nice to aggregate across similar exceptions that are only different in terms of literal values. E.g. in your screenshot, two exceptions were on separate lines, although they were only different by shuffleId- FetchFailed(shuffleId=514/516). 

Screen Shot 2020-07-13 at 2 11 49 PM

**Comment by warrenzhu25** [comment]
> Great improvement
> 
> Would be nice to aggregate across similar exceptions that are only different in terms of literal values. E.g. in your screenshot, two exceptions were on separate lines, although they were only different by shuffleId- FetchFailed(shuffleId=514/516).
> 
> Screen Shot 2020-07-13 at 2 11 49 PM

We could do more advanced aggregation in the future. But for now, I just group by exception type and message, this is easier for user to understand how it work. If we group the ones with similar message together, we need extra place to show difference of these similar messages.

**Comment by warrenzhu25** [comment]
@gengliangwang Could you help take a look?

**Comment by warrenzhu25** [comment]
@dongjoon-hyun @HeartSaVioR could you help take a look?

**Comment by HeartSaVioR** [comment]
cc. @sarutak

**Comment by HeartSaVioR** [comment]
The proposal looks great in general, but I have no idea it's good in UI perspective, so I'd like to defer to @sarutak for reviewing.

**Comment by sarutak** [comment]
@HeartSaVioR Thanks for pinging to me. I'll review within this weekend.
Also cc: @gengliangwang .

**Comment by sarutak** [comment]
@warrenzhu25 Thank you for the contribution. Could you paste screenshots of before applying this change in the description?

**Comment by warrenzhu25** [comment]
> @warrenzhu25 Thank you for the contribution. Could you paste screenshots of before applying this change in the description?

Added in description.

**Comment by sarutak** [comment]
ok to test.

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala [hunk 1]
```diff
@@ -421,6 +420,26 @@ private[spark] class AppStatusStore(
...
+      .flatMap(parseErrorMessage)
+      .groupBy(e => (e.exceptionType, e.message))
+      .map(t => new v1.ExceptionSummary(t._2.head, t._2.length))
+      .toSeq
+      .sortBy(s => (s.count, s.exceptionFailure.exceptionType))(Ordering[(Int, String)].reverse)
+  }
+
+  def parseErrorMessage(errorMessage: String): Option[v1.ExceptionFailure] = {
+    errorMessage.split("\\r?\\n")
+      .find(s => !s.startsWith("\t") && s.contains("Exception:"))
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:436]: * Is it possible to have "\r" and "\t" in the message?
* This code assumes the class name contained in the message ends with `Exception` so this can't cover `Error` e.g, OutOfMemoryError.

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala [hunk 2]
```diff
@@ -421,6 +420,26 @@ private[spark] class AppStatusStore(
     constructTaskDataList(taskDataWrapperIter)
   }
 
+  def exceptionSummary(stageId: Int, attemptId: Int): Seq[v1.ExceptionSummary] = {
+    val tasks = taskList(stageId, attemptId, Int.MaxValue)
+    tasks.filter(t => t.status.equalsIgnoreCase("failed"))
+      .flatMap(t => t.errorMessage)
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:426]: If an exception doesn't have any error message, it can't appear in the summary.

Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:426]: I used flatMap here, if errorMessage is `None`, it won't appear in the summary. Or do you mean empty string?

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala (hunk 1 above)
Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:436]: This errorMessage is actually errorMessage with stacktrace. So it includes:
1. First line contains exception type and error message
2. Stacktrace started at `\t` and end with line break.
I'll rewrite this to only include the first line, then split by ":"

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala (hunk 2 above)
Comment (sarutak) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:426]: Sometimes exceptions can have no error message.
For example, the summary is not shown with the following example.
```
sc.parallelize(1 to 10).foreach(_ => throw new RuntimeException())
```

Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:426]: As the group by condition is exceptionType and message, so I just group them together. Or do you think I should filter them out with `filter(!errorMessage.isEmpty)`?

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 3]
```diff
@@ -223,6 +223,22 @@ private[ui] class StagePage(parent: StagesTab, store: AppStatusStore) extends We
...
+      {if (exceptionSummaryData.nonEmpty) {
+        <span class="collapse-aggregated-exceptionSummaries collapse-table"
+              onClick="collapseTable('collapse-aggregated-exceptionSummaries',
+            'aggregated-exceptionSummaries')">
+          <h4>
+            <span class="collapse-table-arrow arrow-closed"></span>
+            <a>Exception Summary</a>
+          </h4>
+        </span>
+          <div class="aggregated-exceptionSummaries collapsible-table">
```
[the remark this answers is not in the document]
Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:237]: Any suggestion how can I implement the point 2 and 3. I'm not UI expert. I tried to change `arrow-closed` to `arrow-open`, but can't solve point 2

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala (hunk 1 above)
Comment (sarutak) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala]: O.K, I confirmed that stacktrace printed by `printStackTrace` starts with `\t` and each call site is split by `\r\n` on Windows.
But according to the Javadoc, the format, except for the initial line, depends on the implementation.
https://docs.oracle.com/javase/8/docs/api/java/lang/Throwable.html#printStackTrace--

What you want here is the class name and error message, so, can we make this code simpler?

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala (hunk 2 above)
Comment (sarutak) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:426]: I wonder can we show the count somehow even though exceptions have no error message.

**Comment by sarutak** [comment]
This solution groups exceptions by the class name and message so exceptions which have different message will be in different groups. This can cause the summary containing too many entries.
Can we limit the number of entries?

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala (hunk 1 above)
Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala]: Rewrite the logic to handle:
1. exception message is empty
2. Multiple exception types appears will use low level exception type

But still has limitations:
1. As this errorMessage is not exactly `Throwable` errorMessage. It's TaskEndReason.toErrorString, so it could contains something like below:

```
ExecutorLostFailure (executor 389 exited caused by one of the running tasks) Reason: Container from a bad node: container_e1111_1592908128739_15604_01_000741 on host
```
```
FetchFailed(null, shuffleId=516, mapIndex=-1, mapId=-1, reduceId=0, message=
org.apache.spark.shuffle.MetadataFetchFailedException: Missing an output location for shuffle 516
```
2. In the future, we can make TaskEndReason to return exceptionType and message, then we don't parse both from error message.

**Comment by warrenzhu25** [comment]
> This solution groups exceptions by the class name and message so exceptions which have different message will be in different groups. This can cause the summary containing too many entries.
> Can we limit the number of entries?

Return only top 10 entries

**Comment by gengliangwang** [comment]
From the screenshot: https://user-images.githubusercontent.com/1633312/87266114-4bcc7980-c479-11ea-9793-f9e12af645ab.png

It seems that the direction of the arrow besides "Exception Summary" is wrong? Could you update it with the latest screenshot?

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 4]
```diff
@@ -447,6 +464,22 @@ private[ui] class StagePage(parent: StagesTab, store: AppStatusStore) extends We
     </script>
   }
 
+  def exceptionSummaryTable(exceptionSummary: Seq[ExceptionSummary]): Seq[Node] = {
+    val propertyHeader = Seq("Exception", "Message", "Count", "Sample Details")
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:468]: Nit: `Sample Details` => `Sample`

Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala:468]: The reason I use sample is there could be exception which has same message but different stacktrace, but I just show one stacktrace here. I'll change it into details.

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala [hunk 5]
```diff
@@ -421,6 +420,33 @@ private[spark] class AppStatusStore(
     constructTaskDataList(taskDataWrapperIter)
   }
 
+  def exceptionSummary(stageId: Int, attemptId: Int): Seq[v1.ExceptionSummary] = {
+    val tasks = taskList(stageId, attemptId, Int.MaxValue)
```
[the remark this answers is not in the document]
Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:442]: Cached result in store

**Comment by warrenzhu25** [comment]
@sarutak @gengliangwang Could you help check my updates?

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala [hunk 6]
```diff
@@ -421,6 +420,51 @@ private[spark] class AppStatusStore(
     constructTaskDataList(taskDataWrapperIter)
   }
 
+  def exceptionSummary(stageId: Int, attemptId: Int): Seq[v1.ExceptionSummary] = {
+    val (stageData, _) = stageAttempt(stageId, attemptId)
+    val key = Array(stageId, attemptId, stageData.numFailedTasks)
+    asOption(store.read(classOf[CachedExceptionSummary], key))
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:426]: If the exception summary is cached when the stage is in progress, is it possible that we will always see the partial results even after the stage is finished?

Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:426]: I used the number of failed tasks as part of key, if failed tasks increase, it will recompute.

**Comment by warrenzhu25** [comment]
@gengliangwang @sarutak Do you have more comments on this?

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala [hunk 7]
```diff
@@ -421,6 +420,51 @@ private[spark] class AppStatusStore(
...
+  def computeExceptionSummary(stageId: Int, attemptId: Int): Seq[v1.ExceptionSummary] = {
+    val tasks = taskList(stageId, attemptId, Int.MaxValue)
+    tasks.filter(t => t.status.equalsIgnoreCase("failed"))
+      .flatMap(t => t.errorMessage)
+      .flatMap(parseErrorMessage)
+      .groupBy(e => (e.exceptionType, e.message))
+      .map(t => new v1.ExceptionSummary(t._2.head, t._2.length))
+      .toSeq
+      .sortBy(s => (s.count, s.exceptionFailure.exceptionType))(Ordering[(Int, String)].reverse)
+      .take(10)
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:450]: It would be better to have `10` as a named constant.

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (sarutak) [reply core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:442]: `CachedExceptionSummary` seems to never be deleted. This would be a potential cause of memory leak?

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala [hunk 8]
```diff
@@ -421,6 +420,51 @@ private[spark] class AppStatusStore(
...
+      .map(_.exceptionSummary)
+      .getOrElse {
+        val exceptionSummary = computeExceptionSummary(stageId, attemptId)
+        val cachedExceptionSummary = new CachedExceptionSummary(
+          stageId,
+          attemptId,
+          stageData.numFailedTasks,
+          exceptionSummary
+        )
+        store.write(cachedExceptionSummary)
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:436]: `CachedExceptionSummary` seems to never be deleted. This would be a potential cause of memory leak?

**Comment by itskals** [comment]
@warrenzhu25  thanks for the work... 
I think it will add more value if we could add some quantisation fo these errors... So that people understand if it is really worth solving it... One very usefull qualification can be the "time lost" because the time spent earlier by a task is wasted. For example, a partition data was generated by a map task by running for say 1 minute...  the reducers were using this data. But due to some reason, the map data was lost, and the work done by the mapper task( 1 minute worth) had to be redone.
If in my over all application time, if such wasted time is significant, then it is worth doing something to fix it.. else one could just ignore it... 

@HeartSaVioR  @gengliangwang

**Comment by warrenzhu25** [comment]
@gengliangwang @sarutak I rewrote the whole logic to get exception message directly from task failure reason without parsing in most cases. Could you help take a look again?

**Comment by sarutak** [comment]
@Thanks for pinging me. I'll check this weekend.

File: core/src/main/scala/org/apache/spark/status/api/v1/api.scala [hunk 9]
```diff
@@ -393,6 +394,15 @@ class RuntimeInfo private[spark](
     val javaHome: String,
     val scalaVersion: String)
 
+class FailureReason private[spark](
+    val failureType: String,
+    val message: String,
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/status/api/v1/api.scala:399]: The first line of `stackTrace` seems to the same with `message` so it's redundant.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 10]
```diff
@@ -597,6 +598,20 @@ private[spark] class AppStatusListener(
           None
       }
       task.errorMessage = errorMessage
+      task.failureReason = event.reason match {
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:601]: As I mentioned in another place, `message` in FailureReason can be redundant.
If we can remove it, we can just parse error message and build `FailureReason` in `computeFailureSummary` regardless of `event.reason`.

File: core/src/main/scala/org/apache/spark/status/api/v1/api.scala [hunk 11]
```diff
@@ -282,6 +282,7 @@ class TaskData private[spark](
     val speculative: Boolean,
     val accumulatorUpdates: Seq[AccumulableInfo],
     val errorMessage: Option[String] = None,
+    val failureReason: Option[FailureReason] = None,
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/status/api/v1/api.scala:285]: Existing `errorMessage` and newly added `failureReason` are almost the same.
redundant-error-messages
Can we avoid adding `failureReason` by building `FailureReason` in `computeFailureSummary`?

**Review by sarutak — COMMENTED** [review]
I left some comments.
Another concern is that `CachedFailureSummary` is never deleted during an application running.
I think we need to limit the number of entries to be retained.

Also, could you update the screenshot in the description? The current one doesn't reflect the latest change.

cc: @gengliangwang @Dooyoung-Hwang

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 10 above)
Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:601]: Yes, in previous version, I directly passed error message to build failure reason, but I found it hard, complex and error-prone. This due to different failure reason has different format:
1. `ExceptionFailure` is most common and easy one. The message is only exception's message, we can parse it from errorString.
2. `ExecutorLostFailure` errorString has format like `ExecutorLostFailure (executor 72 exited caused by one of the running tasks) Reason: Executor heartbeat timed out after 123153 ms`. We can split the message out using keyword `Reason:`
3. `FetchFailed` errorString has format like `FetchFailed($bmAddressString, shuffleId=$shuffleId, mapId=$mapId, reduceId=$reduceId, message=\n$message\n)`. We need to parse message from this.

In summary, parsing will rely on and couple with specific format of different failure reason. If in the future, the format changed, it's easy to break. So I chose the solution to directly get from TaskEndReason.

Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:601]: @sarutak Any more comments on this?

File: core/src/main/scala/org/apache/spark/status/api/v1/api.scala (hunk 11 above)
Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/status/api/v1/api.scala:285]: Answered in above comments.

File: core/src/main/scala/org/apache/spark/status/api/v1/api.scala (hunk 9 above)
Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/status/api/v1/api.scala:399]: It's not always the same, see below comment for detailed explanation.

**Comment by warrenzhu25** [comment]
@gengliangwang @sarutak @Dooyoung-Hwang I added explanation for the comments. Do you have time to take a look?

**Comment by warrenzhu25** [comment]
@tgravescs Do you have time to take a look?

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala [hunk 12]
```diff
@@ -429,6 +429,37 @@ private[spark] class AppStatusStore(
...
+        store.write(cachedFailureSummary)
+        failureSummary
+      }
+  }
+
+  def computeFailureSummary(stageId: Int, attemptId: Int): Seq[v1.FailureSummary] = {
+    val tasks = taskList(stageId, attemptId, Int.MaxValue)
+    tasks.filter(t => t.status.equalsIgnoreCase("failed"))
+      .flatMap(t => t.failureReason)
+      .groupBy(e => (e.failureType, e.message))
```
Comment (janekdb) [inline core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:456]: Drop use of tuple component accessors by using `values` on the grouped items?

```
.groupBy(e => (e.failureType, e.message))
.values
.map(t => new v1.FailureSummary(t.head, t.length))
```
